### PR TITLE
[Model Runner V2] Add Support for XD-RoPE

### PR DESCRIPTION
**Pull request by santiramos27**
## Summary
- Adds XD-RoPE (variable-dimension RoPE) support to Model Runner V2, enabling models like `HunyuanOCR` that use 3 or 4-dim rotary embeddings
 - Unifies M-RoPE and XD-RoPE into a single RopeState class with one parameterized Triton kernel (`NUM_DIMS: tl.constexpr`), replacing the previous M-RoPE-only MRopeState
 - The kernel always loads and applies delta (non-zero for M-RoPE, zero for XD-RoPE), avoiding a separate code path per variant

 ## Changes

  - `vllm/v1/worker/gpu/mm/rope.py` (new): Unified RopeState class and get_rope_state() factory. The factory validates the model interface (SupportsMRoPE / SupportsXDRoPE) and dispatches based on
  ModelConfig
  - `vllm/v1/worker/gpu/mm/mrope_utils.py` (deleted): Replaced by rope.py
  - `vllm/v1/worker/gpu/model_states/default.py`: Simplified to use get_rope_state() — all M-RoPE/XD-RoPE branching now lives in the factory
  - `vllm/v1/worker/gpu/model_runner.py`: Pass `intermediate_tensors` in `model_inputs` dict directly; non-first PP rank asserts it's not None
  - `vllm/v1/worker/gpu/cudagraph_utils.py`: Add `intermediate_tensors: None` for CUDA graph capture with TODO for PP support (#35162)

## Test Result
- Verified numerical equivalence between V1 and V2 model runners for both XD-RoPE (HunyuanOCR with image input) and M-RoPE (Qwen2-VL-2B-Instruct with image input)
- Identical OCRBench result on V1 vs V2 model runner for HunyuanOCR and Qwen2-VL-2B-Instruct

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/worker/gpu/model_states/default.py [hunk 1]
```diff
@@ -110,21 +131,31 @@ def get_mm_embeddings(
...
-            input_batch.idx_mapping,
-            input_batch.query_start_loc,
-            req_states.prefill_len.gpu,
-            req_states.num_computed_tokens.gpu,
-        )
-        mrope_positions = self.mrope_state.mrope_positions[
-            :, : input_batch.num_tokens_after_padding
-        ]
-        return {"positions": mrope_positions}
+        if self.uses_mrope:
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_states/default.py]: May be good to keep common case at the top, i.e.
```python
if not self.uses_mrope and not self.uses_xdrope:
    # Common case (1D positions).
    return {}

if self.uses_mrope:
    # ...
    return ...

# xdrope logic
return ...

```

File: vllm/v1/worker/gpu/model_states/default.py [hunk 2]
```diff
@@ -60,6 +61,15 @@ def __init__(
                 max_model_len=self.max_model_len,
                 device=self.device,
             )
+        self.uses_xdrope_dim = self.model_config.uses_xdrope_dim
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_states/default.py]: This can be a bool
```suggestion
        self.uses_xdrope = self.model_config.uses_xdrope_dim > 0
```

I actually like using `self.xdrope_state = XDRopeState | None` for this rather than separate bool, but I don't think @WoosukKwon likes that :)

Or, I feel it might be better/clearer to instead have something like a `pe_type` enum (`MROPE`, `XDROPE` or `None`)

File: vllm/v1/worker/gpu/model_runner.py [hunk 3]
```diff
@@ -960,7 +961,6 @@ def execute_model(
             # Update for non-first PP ranks.
             model_inputs["input_ids"] = None
             model_inputs["inputs_embeds"] = None
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py:1003]: we could consider adding this for clarity
```suggestion
            model_inputs["inputs_embeds"] = None
            assert intermediate_tensors is not None
```

**Review by njhill — COMMENTED** [review]
Thanks @santiramos27!

File: vllm/v1/worker/gpu/model_states/default.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/model_states/default.py]: +1. I intentionally put the early exit for the common case at the top. It'd be nice to keep it!

File: vllm/v1/worker/gpu/model_states/default.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/model_states/default.py]: +1. I'm also ok with `if self.xdrope_state is not None`.

File: vllm/v1/worker/gpu/model_states/default.py (hunk 1 above)
Comment (santiramos27) [reply vllm/v1/worker/gpu/model_states/default.py]: Early exit makes sense. updated.

File: vllm/v1/worker/gpu/model_states/default.py (hunk 2 above)
Comment (santiramos27) [reply vllm/v1/worker/gpu/model_states/default.py]: I will change to using `if self.xdrope_state is not None`. From what I can tell, these two PE variants are the only ones getting special treatment in model runner. Should we wait to use enums once we have more PE's to support? It already feels like a good amount of duplication in mrope/xdrope that can probably get consolidated once we have more of them to worry about?

Comment (njhill) [reply vllm/v1/worker/gpu/model_states/default.py]: sure, I just thought it's maybe clearer to have a single enum than two bools which can't both be true

File: vllm/v1/worker/gpu/mm/xdrope_utils.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+import torch
+
+from vllm.model_executor.models.interfaces import SupportsXDRoPE
+from vllm.triton_utils import tl, triton
+from vllm.v1.worker.gpu.buffer_utils import StagedWriteTensor
+
+
+class XDRopeState:
```
Comment (njhill) [inline vllm/v1/worker/gpu/mm/xdrope_utils.py]: Looking at the code now ... it would simplify things to have a shared `RopeState` interface with `init_prefill_positions`, `prepare_positions` and `get_positions(num_tokens)` methods

Comment (santiramos27) [reply vllm/v1/worker/gpu/mm/xdrope_utils.py]: Done. I followed the interface pattern under `model_states/` and did a slight renaming for clarity. Let me know what you think.

File: vllm/v1/worker/gpu/mm/rope/xdrope.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        max_num_reqs: int,
+        max_num_tokens: int,
+        max_model_len: int,
+        device: torch.device,
+    ):
+        self.uses_xdrope_dim = uses_xdrope_dim
+        self.max_num_reqs = max_num_reqs
+        self.max_num_tokens = max_num_tokens
+        self.max_model_len = max_model_len
+        self.device = device
```
Comment (njhill) [inline vllm/v1/worker/gpu/mm/rope/xdrope.py]: Could move these to the superclass

File: vllm/v1/worker/gpu/model_states/default.py [hunk 6]
```diff
@@ -109,31 +118,29 @@ def get_mm_embeddings(
...
         )
-        mrope_positions = self.mrope_state.mrope_positions[
-            :, : input_batch.num_tokens_after_padding
-        ]
-        return {"positions": mrope_positions}
+        return {
+            "positions": self.rope_state.get_positions(
+                input_batch.num_tokens_after_padding
+            )
+        }
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_states/default.py]: nit: I think the original format is cleaner
```suggestion
        positions = self.rope_state.get_positions(
            input_batch.num_tokens_after_padding
        )
        return {"positions": positions}
```

File: vllm/v1/worker/gpu/cudagraph_utils.py [hunk 7]
```diff
@@ -320,6 +320,8 @@ def forward_fn(cg_mode: CUDAGraphMode) -> None:
                     model_inputs = {
                         "input_ids": input_buffers.input_ids[:num_tokens],
                         "positions": input_buffers.positions[:num_tokens],
+                        # NOTE(santinor): PP currently uses eager mode only.
```
Comment (njhill) [inline vllm/v1/worker/gpu/cudagraph_utils.py]: Maybe reference this PR here `https://github.com/vllm-project/vllm/pull/35162`

**Review by njhill — COMMENTED** [review]
Thanks @santiramos27 it looks good to me now.

I can see that there's a natural generalization of both of these, where the mrope uses dim 3 and xdrope uses bias 0. We could have a single kernel and other state logic used by both which would eliminate more of the code. But I have a feeling that @WoosukKwon won't like that idea :-)

File: vllm/v1/worker/gpu/model_states/default.py
Comment (njhill) [inline vllm/v1/worker/gpu/model_states/default.py]: Perhaps we can move this into a `get_rope_state` function in `rope.py` .. which returns `RopeState | None`. It can be passed the model too and check verify it is an instance of `SupportsMRoPE` or `SupportsXDRope` accordingly.

**Review by njhill — COMMENTED** [review]
Thanks @santiramos27 this looks great. Just have one more suggestion.

File: vllm/v1/worker/gpu/model_states/default.py
Comment (santiramos27) [reply vllm/v1/worker/gpu/model_states/default.py]: Done.

File: vllm/v1/worker/gpu/mm/rope.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        self.prefill_delta = UvaBackedTensor(max_num_reqs, dtype=torch.int32)
+
+    def init_prefill_positions(
+        self,
+        req_idx: int,
+        model: nn.Module,
+        prefill_token_ids: list[int],
+        mm_features: list,
+    ) -> None:
+        if self.has_delta:
```
Comment (njhill) [inline vllm/v1/worker/gpu/mm/rope.py:69]: I guess this should be renamed to e.g. `is_mrope` now?

File: vllm/v1/worker/gpu/mm/rope.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+                prefill_token_ids, mm_features
+            )
+
+        for i in range(self.num_dims):
+            pos = prefill_positions[i].tolist()
+            self.prefill_positions.stage_write(self.num_dims * req_idx + i, 0, pos)
+
+    def apply_staged_writes(self) -> None:
+        self.prefill_positions.apply_write()
+        self.prefill_delta.copy_to_uva()
```
Comment (njhill) [inline vllm/v1/worker/gpu/mm/rope.py]: probably still worth has_delta (or is_mrope) check here?

File: vllm/v1/worker/gpu/mm/rope.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        assert isinstance(model, SupportsMRoPE)
+        return RopeState(
+            num_dims=3,
+            has_delta=True,
+            max_num_reqs=max_num_reqs,
+            max_num_tokens=max_num_tokens,
+            max_model_len=max_model_len,
+            device=device,
+        )
+    elif model_config.uses_xdrope_dim > 0:
```
Comment (njhill) [inline vllm/v1/worker/gpu/mm/rope.py]: nit
```suggestion
    if model_config.uses_xdrope_dim > 0:
```

**Review by njhill — COMMENTED** [review]
Thanks @santiramos27 just couple of last minor things!

File: vllm/v1/worker/gpu/mm/rope.py (hunk 10 above)
Comment (santiramos27) [reply vllm/v1/worker/gpu/mm/rope.py]: mutually exclusive but i forgot i added early return here :')

File: vllm/v1/worker/gpu/mm/rope.py (hunk 9 above)
Comment (santiramos27) [reply vllm/v1/worker/gpu/mm/rope.py]: You're right -- for xdrope the UVA tensor is zero-initialized and we never write to it, so this copy is unnecessary

File: vllm/v1/worker/gpu/mm/rope.py (hunk 10 above)
Comment (njhill) [reply vllm/v1/worker/gpu/mm/rope.py]: yeah it's not incorrect maybe more a code style preference :)

File: vllm/v1/worker/gpu/mm/rope.py (hunk 8 above)
Comment (santiramos27) [reply vllm/v1/worker/gpu/mm/rope.py:69]: Since `get_rope_state()` already handles M-RoPE vs XD-RoPE dispatch, has_delta feels more natural here — it describes the behavioral difference rather than leaking model-type identity into the unified RopeState (ie. mrope and xdrope are instances of RopeState). Wdyt?

**Review by njhill — APPROVED** [review]
Thanks @santiramos27 for the nice work!
